File: Coding/golang/bloomfilter/bloomfilter.go
package main

import "fmt"

// bitSet is a minimal, growable bit set used as the filter backing store
// (a port of java.util.BitSet, which also grows on demand).
type bitSet struct {
	words []uint64
}

// set turns on bit i, growing the backing store as needed.
func (b *bitSet) set(i int) {
	w := i >> 6
	for w >= len(b.words) {
		b.words = append(b.words, 0)
	}
	b.words[w] |= 1 << uint(i&63)
}

// get reports whether bit i is set.
func (b *bitSet) get(i int) bool {
	w := i >> 6
	if w >= len(b.words) {
		return false
	}
	return b.words[w]&(1<<uint(i&63)) != 0
}

// BloomFilter is a probabilistic set-membership structure with no false
// negatives (but possible false positives).
type BloomFilter struct {
	filter    *bitSet
	hashCount int
}

// newBloomFilter builds a filter using the given number of hash functions and
// an existing bit set.
func newBloomFilter(hashes int, filter *bitSet) *BloomFilter {
	return &BloomFilter{filter: filter, hashCount: hashes}
}

// add inserts key by setting each of its hash-bucket bits. O(hashCount).
func (bf *BloomFilter) add(key string) {
	for _, bucketIndex := range getHashBuckets(key, bf.hashCount) {
		bf.filter.set(bucketIndex)
	}
}

// isPresent reports whether key may be present. A false result is definitive;
// a true result may be a false positive. O(hashCount).
func (bf *BloomFilter) isPresent(key string) bool {
	for _, bucketIndex := range getHashBuckets(key, bf.hashCount) {
		if !bf.filter.get(bucketIndex) {
			return false
		}
	}
	return true
}

// getHashBuckets is a faithful port of the original Java method, which is a
// stub that returns hashCount zero-valued buckets (real hashing was never
// implemented there). Consequently every key maps to bucket 0.
func getHashBuckets(key string, hashCount int) []int {
	result := make([]int, hashCount)
	return result
}

func main() {
	bf := newBloomFilter(3, &bitSet{})
	bf.add("hello")
	// Note: because getHashBuckets is the original stub (all-zero buckets),
	// any key reports present once any key has been added.
	fmt.Println("isPresent(hello):", bf.isPresent("hello"))
	fmt.Println("isPresent(world):", bf.isPresent("world"))
}

File: Coding/python/bloom_filter.py
"""Bloom filter ported from BloomFilter.java (original Java package code.ds).

The Java BitSet is represented here as a Python set of set-bit indices.

NOTE: get_hash_buckets is a *faithful* stub -- the reference returns an
all-zero array of length hash_count, so every key maps to bit 0. As a result,
once any key is added, is_present() returns True for every key (false positives
by design). The class structure and add / query logic are preserved exactly.
"""


class BloomFilter:

    def __init__(self, hashes, filter_bits=None):
        self.hash_count = hashes
        # Java BitSet -> a set holding the indices of the bits that are set.
        self.filter_ = filter_bits if filter_bits is not None else set()

    def add(self, key):
        for bucket_index in self.get_hash_buckets(key, self.hash_count):
            self.filter_.add(bucket_index)

    def is_present(self, key):
        for bucket_index in self.get_hash_buckets(key, self.hash_count):
            if bucket_index not in self.filter_:
                return False
        return True

    @staticmethod
    def get_hash_buckets(key, hash_count):
        # Faithful stub: the reference returns a zero-filled array.
        return [0] * hash_count


if __name__ == "__main__":
    bf = BloomFilter(hashes=3)
    print("is_present('apple') before add ->", bf.is_present("apple"))  # False
    bf.add("apple")
    print("is_present('apple') after add  ->", bf.is_present("apple"))  # True
    # Because of the stub every key maps to bit 0 -> false positive by design:
    print("is_present('banana')           ->", bf.is_present("banana"))  # True
